axon-server-connector/src/main/java/org/axonframework/axonserver/connector/command/AxonServerCommandBusConnector.java

```diff
@@ -53,26 +55,49 @@ public class AxonServerCommandBusConnector implements CommandBusConnector {
     private static final Logger logger = LoggerFactory.getLogger(AxonServerCommandBusConnector.class);
 
     private final AxonServerConnection connection;
-    private CommandBusConnector.Handler incomingHandler;
+    private final String clientId;
+    private final String componentName;
+
+    private Handler incomingHandler;
     private final Map<QualifiedName, Registration> subscriptions = new ConcurrentHashMap<>();
+    private final ShutdownLatch shutdownLatch = new ShutdownLatch();
 
     /**
      * Creates a new {@code AxonServerConnector} that communicate with Axon Server using the provided
      * {@code connection}.
      *
-     * @param connection The {@code AxonServerConnection} to communicate to Axon Server with.
+     * @param connection    The {@code AxonServerConnection} to communicate to Axon Server with.
+     * @param configuration The Axon Server configuration, used to retrieve (e.g.) the
+     *                      {@link AxonServerConfiguration#getClientId()} to be set when
+     *                      {@link #dispatch(CommandMessage, ProcessingContext) dispatching} commands.
      */
-    public AxonServerCommandBusConnector(@Nonnull AxonServerConnection connection) {
+    public AxonServerCommandBusConnector(@Nonnull AxonServerConnection connection,
+                                         @Nonnull AxonServerConfiguration configuration) {
         this.connection = Objects.requireNonNull(connection, "The AxonServerConnection must not be null.");
+        Objects.requireNonNull(configuration, "The AxonServerConfiguration must not be null.");
+        this.clientId = configuration.getClientId();
+        this.componentName = configuration.getComponentName();
+    }
+
+    /**
+     * Starts the Axon Server {@link CommandBusConnector} implementation.
+     */
+    public void start() {
+        shutdownLatch.initialize();
+        logger.trace("The AxonServerCommandBusConnector started.");
     }
 
     @Nonnull
     @Override
     public CompletableFuture<CommandResultMessage<?>> dispatch(@Nonnull CommandMessage<?> command,
                                                                @Nullable ProcessingContext processingContext) {
-        return connection.commandChannel()
-                         .sendCommand(CommandConverter.convertCommandMessage(command))
-                         .thenCompose(CommandConverter::convertCommandResponse);
+        shutdownLatch.ifShuttingDown("Cannot dispatch new commands as this bus is being shutdown");
+        try (ShutdownLatch.ActivityHandle commandInTransit = shutdownLatch.registerActivity()) {
```

Comment:
Wow, nice usage - I've learned something new. I haven't known this API with `registerActivity`.

Reply:
We made it specifically when introducing the graceful start up and shutdown feature in Axon Framework 4.3! Currently only used for the two-staged shutdown components, though.